modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -202,21 +215,85 @@ public CompletableFuture<Void> finish(
     ) {
         assert groups != null && !groups.isEmpty();
 
-        HybridTimestamp commitTimestamp = commit ? clock.now() : null;
-
-        TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
-                .txId(txId)
-                .timestampLong(clock.nowLong())
-                .groupId(commitPartition)
-                .groups(groups)
-                .commit(commit)
-                .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
-                .term(term)
-                .build();
-
-        return replicaService.invoke(recipientNode, req)
-                // TODO: IGNITE-20033 TestOnly code, let's consider using Txn state map instead of states.
-                .thenRun(() -> changeState(txId, PENDING, commit ? COMMITED : ABORTED));
+        Function<Void, CompletableFuture<Void>> clo = ignored -> {
+            HybridTimestamp commitTimestamp = commit ? clock.now() : null;
+
+            TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
+                    .txId(txId)
+                    .timestampLong(clock.nowLong())
+                    .groupId(commitPartition)
+                    .groups(groups)
+                    .commit(commit)
+                    .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
+                    .term(term)
+                    .build();
+
+            return replicaService.invoke(recipientNode, req)
+                    // TODO: IGNITE-20033 TestOnly code, let's consider using Txn state map instead of states.
+                    .thenRun(() -> changeState(txId, PENDING, commit ? COMMITED : ABORTED));
+        };
+
+        if (!commit) {
+            AtomicReference<CompletableFuture<Void>> ref = new AtomicReference<>();
+            TxContext tuple = txCtxMap.compute(txId, (uuid, tuple0) -> {
+                if (tuple0 == null) {
+                    tuple0 = new TxContext(); // No writes enlisted.
```

Comment:
Let's be consistent with the naming. enlistment and in-flight operations have different meaning. It's minor of course.

Reply:
The naming is correct here - inflighs are caused by enlisted writes.